quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -78,10 +232,178 @@ impl Builder {
+    let mut default_sender = DefaultSender::builder()
+        .with_capacity(2)
+        .with_connection_info(conn_info)
+        .build()
+        .unwrap();
+    let datagram_0 = bytes::Bytes::from_static(&[1, 2, 3]);
+    let datagram_1 = bytes::Bytes::from_static(&[4, 5, 6]);
+    let datagram_2 = bytes::Bytes::from_static(&[7, 8, 9]);
+
+    let (waker, _counter) = new_count_waker();
+    let mut cx = Context::from_waker(&waker);
+
+    assert_eq!(
+        default_sender.poll_send_datagram(datagram_0, &mut cx),
+        Poll::Ready(Ok(()))
+    );
+
+    assert_eq!(
+        default_sender.poll_send_datagram(datagram_1, &mut cx),
+        Poll::Ready(Ok(()))
+    );
+
+    // Waker has not been set up yet because queue is not yet at capacity
+    assert!(default_sender.waker.is_none());
+    assert_eq!(
+        default_sender.poll_send_datagram(datagram_2, &mut cx),
+        Poll::Pending
+    );
+
+    // Since queue is at capacity default_sender is now storing a waker that will
+    // alert when the queue has more space
+    assert!(default_sender.waker.is_some());
+
+    // Check that the first two datagrams are on the queue
+    let first = default_sender.queue.pop_front().unwrap();
+    assert_eq!(first.data[..], [1, 2, 3]);
+    let second = default_sender.queue.pop_front().unwrap();
+    assert_eq!(second.data[..], [4, 5, 6]);
+}
+
+#[test]
+fn retain_datagrams() {
+    let conn_info = ConnectionInfo {
+        max_datagram_payload: 100,
+    };
+    // Create a default sender queue
+    let mut default_sender = DefaultSender::builder()
+        .with_capacity(3)
+        .with_connection_info(conn_info)
+        .build()
+        .unwrap();
+    let datagram_0 = bytes::Bytes::from_static(&[1, 2, 3]);
+    let datagram_1 = bytes::Bytes::from_static(&[4, 5, 6]);
+    let datagram_2 = bytes::Bytes::from_static(&[7, 8, 9]);
+    assert_eq!(default_sender.send_datagram(datagram_0), Ok(None));
+    assert_eq!(default_sender.send_datagram(datagram_1), Ok(None));
+    assert_eq!(default_sender.send_datagram(datagram_2), Ok(None));
+
+    // Keep only the third datagram
+    default_sender.retain_datagrams(|datagram| datagram.data[..] == [7, 8, 9]);
+    let first = default_sender.queue.pop_front().unwrap();
+    assert_eq!(first.data[..], [7, 8, 9]);
+    assert!(default_sender.queue.is_empty());
+}
+
+#[test]
+fn record_capacity_stats() {
+    // Create a default sender queue
+    let mut default_sender = DefaultSender::builder().build().unwrap();
+    default_sender.record_capacity_stats(100);
+    default_sender.record_capacity_stats(100);
+    default_sender.record_capacity_stats(200);
+    default_sender.record_capacity_stats(500);
+
+    assert_eq!(default_sender.max_packet_space(), 500);
+    assert_eq!(default_sender.min_packet_space(), 100);
+
+    // There's not a great test for a correct weighted average, but we
+    // can at least check our output isn't zero
+    assert!(default_sender.smoothed_packet_space() > 0.0);
```

Comment:
Why not just hard code what it is? Or at the very least you can do a snapshot

```suggestion
    insta::assert_debug_snapshot!(default_sender.smoothed_packet_space());
```